abi-versions/vNEXT/README.md

```diff
@@ -0,0 +1,1471 @@
+
+Return value (`next_action`) instructs the host environment what action to take:
+- `Continue` means that HTTP trailers should be forwarded downstream.
+- `EndStream` means that HTTP trailers should be forwarded downstream and that
+  the underlying stream should be closed afterwards.
+- `Done` means that HTTP trailers should be forwarded downstream and that this
+  extension shouldn't be called again for this HTTP response.
+- `Pause` instructs the host environment to pause processing until
+  `proxy_resume` is called.
+- `Close` means that the underlying stream should be closed immediately.
+
+#### `proxy_on_http_response_metadata`
+
+* params:
+  - `i32 (uint32_t) stream_id`
+  - `i32 (size_t) num_elements`
+* returns:
+  - `i32 (proxy_action_t) next_action`
+
+Called for each HTTP/2 METADATA frame for stream `stream_id` received from
+upstream.
+
+Metadata (`num_elements` entries) can be retrieved using `proxy_get_map_values`.
+
+Return value (`next_action`) instructs the host environment what action to take:
+- `Continue` means that HTTP/2 METADATA elements should be forwarded downstream.
+- `EndStream` means that HTTP/2 METADATA elements should be forwarded downstream
+  and that the underlying stream should be closed afterwards.
+- `Done` means that HTTP/2 METADATA elements should be forwarded downstream and
+  that this extension shouldn't be called again for this HTTP response.
+- `Pause` instructs the host environment to pause processing until
+  `proxy_resume` is called.
+- `Close` means that the underlying stream should be closed immediately.
+
+### Functions exposed by the host environment
+
+#### `proxy_send_http_response`
+
+* params:
+  - `i32 (uint32_t) response_code`
+  - `i32 (const char*) response_code_details_data`
+  - `i32 (size_t) response_code_details_size`
+  - `i32 (iovec_array_t) response_body_data`
+  - `i32 (size_t) response_body_size`
+  - `i32 (iovec_array_t[]) additional_headers_map_data`
+  - `i32 (size_t) additional_headers_size`
+  - `i32 (uint32_t) grpc_status`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Sends HTTP response to the client, without forwarding request to the upstream.
+
+Returns `Ok` on success.
+
+Returns `BadArgument` for invalid `response_code` and/or `grpc_status`.
+
+Returns `InvalidMemoryAccess` when `response_code_details_data`,
+`response_code_details_size`, `response_body_data`, `response_body_size`,
+`additional_headers_map_data` and/or `additional_headers_size` point to invalid
+memory address.
+
+#### `proxy_resume_http_stream`
+
+* params:
+  - `i32 (proxy_stream_type_t) stream_type`
+* returns:
+  - `i32 (proxy_result_t) call_result`
+
+Resume processing of paused HTTP request or response (`stream_type`) in current
+context.
+
+Returns `Ok` on success.
+
+Returns `BadArgument` for unknown `stream_type`.
+
+Returns `InvalidOperation` when stream wasn't paused.
+
+#### `proxy_close_http_stream`
```

Comment:
Similar comment about if we have this why do we need the special return codes from the filter callbacks. Seems like it would be simpler to reduce the return codes there and just have people call this before they stop iteration or whatever, though I don't feel strongly about it.